ci/travis/ci.sh

```diff
@@ -38,21 +38,187 @@ should_run_job() {
+    esac
+    export TRAVIS_PULL_REQUEST
+  fi
+
+  local definitions=() paths=()
+
+  # Environment variables for all platforms
+  definitions+=(
+    PYTHON3_BIN_PATH=python
+    GOROOT="${HOME}/go"
+    GOPATH="${HOME}/go_dir"
+  )
+  # Environment variables for Windows
+  test "${OSTYPE}" != msys || definitions+=(
+    USE_CLANG_CL=1
+  )
+
+  # NOTE: Modifying PATH invalidates Bazel's cache! Do not add to PATH unnecessarily.
+  # PATH for all platforms
+  paths+=(
+    "${HOME}/miniconda/bin"
+  )
+  # PATH for Windows
+  test "${OSTYPE}" != msys || paths+=(
+    "${HOME}/miniconda/bin/Scripts"
+  )
+
+  export "${definitions[@]}" 2> /dev/null
+
+  prepend_PATH "${paths[@]}"
+}
+
+# Prepends each argument to PATH if it isn't already in the PATH.
+prepend_PATH() {
   { local set_x="${-//[^x]/}"; } 2> /dev/null  # save set -x to suppress noise
   set +x
-  local to_add="$HOME/miniconda/bin" old_path=":${PATH}:"
-  if [ "${old_path}" = "${old_path##*:${to_add}:*}" ]; then
-    PATH="${to_add}:$PATH"
+  local old="${PATH}"
+  if [ "${OSTYPE}" = msys ]; then
+    PATH="${PATH// :/:}"  # HACK: Work around https://github.com/actions/virtual-environments/issues/635#issue-589318442
+  fi
+
+  local to_add
+  for to_add in "$@"; do
+    case ":${PATH}:" in
+      *:"${to_add}":*) ;;  # Do nothing; already in PATH
+      *) PATH="${to_add}:${PATH}" ;;  # Prepend to PATH
+    esac
+  done
+
+  if [ "${old}" != "${PATH}" ]; then
+    echo "PATH=${PATH}"
   fi
   test -z "${set_x}" || set -x  # restore set -x
+}
+
+test_python() {
   if [ "${OSTYPE}" = msys ]; then
-    PATH="${PATH// :/:}"  # HACK: Work around https://github.com/actions/virtual-environments/issues/635#issue-589318442
+    # Windows -- most tests won't work yet; just do the ones we know work
+    PYTHONPATH=python python -m pytest --durations=5 --timeout=300 python/ray/tests/test_mini.py
+  fi
+}
+
+build_sphinx_docs() {
+  (
+    cd "${WORKSPACE_DIR}"/doc
+    if [ "${OSTYPE}" = msys ]; then
+      echo "WARNING: Documentation not built on Windows due to currently-unresolved issues"
+    else
+      sphinx-build -q -W -E -T -b html source _build/html
+    fi
+  )
+}
+
+lint_readme() {
+  (
+    cd "${WORKSPACE_DIR}"/python
+    python setup.py check --restructuredtext --strict --metadata
+  )
+}
+
+lint_python() {
+  # ignore dict vs {} (C408), others are defaults
+  command -V python
+  python -m flake8 --inline-quotes '"' --no-avoid-escape --exclude=python/ray/core/generated/,streaming/python/generated,doc/source/conf.py,python/ray/cloudpickle/,python/ray/thirdparty_files --ignore=C408,E121,E123,E126,E226,E24,E704,W503,W504,W605
+  "${ROOT_DIR}"/format.sh --all
+}
+
+lint_bazel() {
+  local platform="" architecture="" suffix=".tar.gz"
+  case "${OSTYPE}" in
+    linux*) platform=linux;;
+    darwin*) platform=darwin;;
+    msys*) platform=windows; suffix=".zip";;
+  esac
+  case "${HOSTTYPE}" in
+    x86_64) architecture=amd64;;
+    i?86*) architecture=i386;;
```

Comment:
Why do we need to support that?

Reply:
I guess we don't; I just added the `case` to make it clear there's a translation from `HOSTTYPE` and then this seemed like a cheap line to add afterwards. I can remove it.